### PR TITLE
Honor lambda states (optional now, rolled out everywhere in September)

### PR DESCRIPTION
**Pull request by astuyve**
- feat: Honor lambda state in instrument method
- feat: double escape quotes to pass snapshots

### What and why?

This PR verifies that the function can be updated, which according to this [post](https://aws.amazon.com/blogs/compute/coming-soon-expansion-of-aws-lambda-states-to-all-functions/) is when the `State` is `Active` and `LastUpdateStatus` is `Successful.

### How?

When we fetch the function configuration, this is attribute is returned. Because this command is likely ran in CI, immediately after updating a function, we will try 3 extra times if the first state is not valid, with exponential backoff between attempts.

I believe _all_ function configurations return `State` and `LastUpdateStatus`, but I've included a guard clause in cause. To be removed 1 Oct 2021, as per comment and schedule.

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)

### CODE REVIEW

File: src/commands/lambda/function.ts [hunk 1]
```diff
@@ -40,6 +40,46 @@ export interface InstrumentationSettings {
...
+
+/**
+ * Waits for n ms
+ * @param ms
+ * @returns
+ */
+const wait = (ms: number): Promise<void> => new Promise((res) => setTimeout(res, ms))
+
+/**
+ *
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/function.ts:53]: Is this doc comment just here for linting reasons? Seems redundant with the types?

File: src/commands/lambda/__tests__/instrument.test.ts [hunk 2]
```diff
@@ -235,6 +235,42 @@ describe('lambda', () => {
                                         `)
       })
 
+      test('aborts if a function is not in an Active state with LastUpdateStatus Successful', async () => {
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/__tests__/instrument.test.ts:238]: Nice test

File: src/commands/lambda/function.ts [hunk 3]
```diff
@@ -40,6 +40,46 @@ export interface InstrumentationSettings {
...
+ * @param config
+ * @param functionArn
+ * @param attempts
+ * @returns bool
+ */
+const checkLambdaState = async (
```
Comment (agocs) [inline src/commands/lambda/function.ts:59]: Maybe rename to `isLambdaActive` or `isUpdatable`. `checkLambdaState` isn't a clear name, especially given that it returns a boolean.

File: src/commands/lambda/function.ts [hunk 4]
```diff
@@ -40,6 +40,46 @@ export interface InstrumentationSettings {
...
+): Promise<boolean> => {
+  // TODO remove 1 Oct 2021 https://aws.amazon.com/blogs/compute/tracking-the-state-of-lambda-functions/
+  if (!config.State || !config.LastUpdateStatus) {
+    return true
+  }
+  if (config.LastUpdateStatus === 'Successful' && config.State === 'Active') {
+    return true
+  }
+  if (config.State === 'Pending' && attempts <= MAX_LAMBDA_STATE_CHECKS) {
+    await wait(2 ** attempts * 1000)
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/function.ts:73]: At a glance I wasn't sure what the order of operations between ** and * is. Maybe add parenthesis? Or use  Math.pow over **.

**Review by DarcyRaynerDD — APPROVED** [review]
LGTM, left some nits

**Review by agocs — APPROVED** [review]
Looks good! Just a nitpick about the name `checkLambdaState`

File: src/commands/lambda/function.ts (hunk 1 above)
Comment (astuyve) [reply src/commands/lambda/function.ts:53]: Bad editor config, need to remove jsdoc for `.ts`

**Pull request merged**